clang/include/clang/AST/Type.h

```diff
@@ -2686,32 +2683,55 @@ class ParenType : public Type, public llvm::FoldingSetNode {
   static bool classof(const Type *T) { return T->getTypeClass() == Paren; }
 };
 
+/// This class augments a type with a pointer interpretation.
+template <class T>
+class PointerInterpretationTrait {
+protected:
+  PointerInterpretationTrait() = default;
+
+public:
+  bool isCHERICapability() const {
+    return getPointerInterpretation() == PIK_Capability;
+  }
+
+  PointerInterpretationKind getPointerInterpretation() const {
+    return static_cast<const T *>(this)->getPointerInterpretationImpl();
+  }
+};
+
 /// PointerType - C99 6.7.5.1 - Pointer Declarators.
-class PointerType : public Type, public llvm::FoldingSetNode {
+class PointerType : public Type,
+                    public PointerInterpretationTrait<PointerType>,
+                    public llvm::FoldingSetNode {
   friend class ASTContext; // ASTContext creates these.
+  friend class PointerInterpretationTrait<PointerType>;
 
   QualType PointeeType;
-  bool IsCHERICapability : 1;
+  PointerInterpretationKind PIK;
```

Comment:
Since we are changing this line, maybe we should move it to the base class bitfields?